Working on cross-compiling

Nico-Curti opened the pull request:
Add some rules to create nocs shared and static libraries and add a pretty layout.
Fix many warnings and some (mini) errors.
Add CMakeFile
Working on Windows compatibility...

Nico-Curti committed:
pretty makefile and add build libs

Nico-Curti committed:
Add Cmake, fix many warnings and errors, add more CI

Nico-Curti committed:
miss cd

Nico-Curti committed:
fix typo test

Nico-Curti committed:
minor fix and add more CI

Nico-Curti committed:
minor fix for CI

Nico-Curti committed:
fix names

Nico-Curti committed:
minor fix and change std

cenit commented on .travis.yml:
```diff
@@ -287,5 +287,5 @@ install:
-  - cd bin && ./testing && cd ..
-  - cd bin && ./testing && cd ..
+  - cd bin && ./Test && cd ..
+  - cd bin && ./Test && cd ..
```
sai vero che lo esegui due volte e basta? :)

Nico-Curti replied:
Sì sì sto solo facendo delle prove…
Adesso infatti ho messo a posto anche quello (spero)

Nico-Curti committed:
minor fix

Nico-Curti committed:
try to fix std clang

Nico-Curti committed:
fix libstd

Nico-Curti committed:
add build script and minor fix on CMake

Nico-Curti committed:
try to fix clang 7

Commit:
Replace std::tuple initializer list with proper explicit constuctor

Apparently, with older versions of g++, the constructor for std::tuple is explicit.
In order to maximize compatibility, we need to switch to this notation.
This breaks the "calligraphy" of the code a bit, but it's DEFINETLY for the best.

Commit:
Merge branch 'master' of https://github.com/Nico-Curti/nocs

carlidel commented:
Dovrei aver sistemato quell'errore di compilazione su g++-5... ma non so esattamente se/come/dove ho finito per pushare i cambiamenti 😅

carlidel commented:
Ok, vedo che stanno allineati qua e che ho praticamente invaso il fork di Nico senza nemmeno chiedere permesso... chiedo venia per l'intrusione

Nico-Curti committed:
first attempt to remove passe_par_tout

Nico-Curti committed:
enlarge readme and remove some allow_failures from CI

Nico-Curti committed:
fix of previous fix (??)

Nico-Curti committed:
add pthread dependency. graphics still not work

Nico-Curti committed:
msvc works! graphic still not work!

Commit:
Add first (messy) working graphics

We now have a first working verson with the new graphics... but it's still necessary to clean a lot of stuff and check for stability problems (it seems that sometimes the code randomly freezes or crashes).

carlidel commented:
Sono finalmente riuscito a prendermi un po' di tempo per imparare due righe di GLUT e far andare una mezza grafica decente su NOCS... ora vedo di capire se posso togliere un po' di robe vestigiali rimaste dalla vecchia versione basata su passe_par_tout.

Commit:
Bit of cleaning, but still crashes

carlidel commented:
A quanto leggo in giro, GLUT di base non è indicato per nessun tipo di lavoro in multithreading... tant'è che tutti i forum che trovo dicono che lanciare `glutMainLoop()` al difuori del main thread risulta essere molto pericoloso.

Vedo se trovo qualche metodo rapido per rendere vivibile questa cosa...

Commit:
Add working graphics with major refactoring

This proper multithreaded implementation places the whole GLUT istance in a completely separate thread with proper mutex and atomic communication.

There are still some TODOs:
> reimplement properly wait_click() and wait_enter(), they now don't work.
> improve the interface for the graphical methods, right now is kinda basic.
> evaluate a proper framerate timing, since now it will inevitably skip more than some frames if the simulation is faster than the graphical rendering.

Nico-Curti committed:
minor minor minor change

Commit:
Fix wait_click, remove wait_enter, remove useless thread sleep

The main idea behind the implementation of wait_click was just plain wrong... You never ever interfere with the flow of the GLUT thread, instead you have to freeze the execution on the main thread.

This way, the callback functions actually work and mouse click is detected.

Commit:
Remove window :: clear()

It's not needed anymore

Commit:
Assign different color to bumpers

Little better but in the future it might be ideal to study an actual system for custom color assignment.

Commit:
Merge flush and draw methods

As for now... I'd say we don't need these two methods to be detached.
Maybe in the future? When the tag system will be more documented? We'll see.
